pageserver/src/consumption_metrics/upload.rs

```diff
@@ -0,0 +1,264 @@
+        uploaded,
+        failed,
+        elapsed_ms = elapsed.as_millis(),
+        "done sending metrics"
+    );
+
+    Ok(())
+}
+
+// The return type is quite ugly, but we gain testability in isolation
+fn serialize_in_chunks<'a, F>(
+    chunk_size: usize,
+    input: &'a [RawMetric],
+    factory: F,
+) -> impl Iterator<Item = Result<(&'a [RawMetric], bytes::Bytes), serde_json::Error>> + 'a
+where
+    F: KeyGen<'a> + 'a,
+{
+    use bytes::BufMut;
+
+    // write to a BytesMut so that we can cheaply clone the frozen Bytes for retries
+    let mut buffer = bytes::BytesMut::new();
+    let mut chunks = input.chunks(chunk_size);
+    let mut events = Vec::new();
+
+    std::iter::from_fn(move || {
+        let chunk = chunks.next()?;
+
+        if !events.is_empty() {
+            assert_eq!(events.len(), chunk_size);
+            events
+                .iter_mut()
+                .zip(chunk.iter())
+                .for_each(|(slot, raw_metric)| {
+                    raw_metric.update_in_place(slot, factory.generate())
+                });
+        } else {
+            events.extend(
+                chunk
+                    .iter()
+                    .map(|raw_metric| raw_metric.as_event(factory.generate())),
+            );
+        }
+
+        let res = serde_json::to_writer(
+            (&mut buffer).writer(),
+            &EventChunk {
+                events: (&events[..chunk.len()]).into(),
+            },
+        );
+
+        match res {
+            Ok(()) => Some(Ok((chunk, buffer.split().freeze()))),
+            Err(e) => Some(Err(e)),
+        }
+    })
+}
+
+trait KeyGen<'a>: Copy {
+    fn generate(&self) -> IdempotencyKey<'a>;
+}
+
+impl<'a> KeyGen<'a> for &'a str {
+    fn generate(&self) -> IdempotencyKey<'a> {
+        IdempotencyKey::generate(self)
+    }
+}
+
+enum UploadError {
+    Rejected(reqwest::StatusCode),
+    Reqwest(reqwest::Error),
+    Cancelled,
+}
+
+impl std::fmt::Debug for UploadError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        // use same impl because backoff::retry will log this using both
+        std::fmt::Display::fmt(self, f)
+    }
+}
+
+impl std::fmt::Display for UploadError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        use UploadError::*;
+
+        match self {
+            Rejected(code) => write!(f, "server rejected the metrics with {code}"),
+            Reqwest(e) => write!(f, "request failed: {e}"),
+            Cancelled => write!(f, "cancelled"),
+        }
+    }
+}
+
+impl UploadError {
+    fn is_reject(&self) -> bool {
+        matches!(self, UploadError::Rejected(_))
+    }
+}
+
+async fn upload(
+    client: &reqwest::Client,
+    metric_collection_endpoint: &reqwest::Url,
+    body: bytes::Bytes,
+    cancel: &CancellationToken,
+) -> Result<(), UploadError> {
+    if cancel.is_cancelled() {
+        // retry does not drop the request when cancel comes, so pre-check it.
```

Comment:
Not sure what this means: `backoff::retry` does check the cancellation token at the start of its loop each time, so if the cancellation token is set then we will never call the inner function and drop out immediately.  Isn't that the same as what this path is doing?

Reply:
I was thinking `backoff::retry` would do no such thing, but happy to remove it.

Reply:
I had misunderstood, 00a9baa91